File: src/downloads_organizer/semantic_text.py
from __future__ import annotations

from .models import IndexedFile

SEMANTIC_TEXT_VERSION = "3"


def _sample(value: str, budget: int) -> str:
    value = " ".join(value.split())
    if len(value) <= budget:
        return value
    if budget < 12:
        return value[:budget]
    separator_budget = len(" … ") * 2
    part = max(1, (budget - separator_budget) // 3)
    middle = max(0, len(value) // 2 - part // 2)
    return " … ".join((value[:part], value[middle:middle + part], value[-part:]))


def build_semantic_text(file: IndexedFile, *, max_chars: int = 2400) -> str:
    """Build a short, stable representation for embedding and translation.

    Metadata gets a reserved prefix and the extracted body is sampled from the
    front, middle, and tail.  The function never returns more than max_chars.
    """
    if max_chars < 256:
        raise ValueError("semantic text budget must be at least 256 characters")

    sections: list[str] = []
    seen: set[str] = set()
    for label, value in (
        ("Title", file.title or file.path.stem),
        ("Summary", file.summary),
        ("Keywords", ", ".join(file.keywords)),
    ):
        cleaned = " ".join(value.split())
        key = cleaned.casefold()
        if cleaned and key not in seen:
            sections.append(f"{label}: {cleaned}")
            seen.add(key)

    prefix = "\n".join(sections)
    body_budget = max(0, max_chars - len(prefix) - (2 if prefix else 0))
    body = _sample(file.text, body_budget) if body_budget else ""
    result = "\n\n".join(part for part in (prefix, body) if part)
    return result[:max_chars]


def semantic_language(file: IndexedFile) -> str:
    """Choose a stable native embedding space from human-facing metadata.

    Code-heavy course notes frequently confuse automatic language detection.
    Titles and summaries are a better signal for the document's prose language.
    """
    sample = " ".join((file.title or file.path.stem, file.summary[:600]))
    if any("\uac00" <= char <= "\ud7af" for char in sample):
        return "ko"
    if any("\u3040" <= char <= "\u30ff" for char in sample):
        return "ja"
    cjk = sum("\u4e00" <= char <= "\u9fff" for char in sample)
    latin = sum(char.isascii() and char.isalpha() for char in sample)
    if cjk >= 2 and cjk / max(1, cjk + latin) >= 0.12:
        return "zh-Hans"
    return "en"


def semantic_cache_version(encoder_version: str) -> str:
    return f"{encoder_version};semantic-text:{SEMANTIC_TEXT_VERSION}"

File: Sources/TopicTidyCore/Semantic/SemanticText.swift
import Foundation

public enum SemanticText {
    public static let version = "3"
    public static let viewVersion = "1"
    public static let views = ["identity", "overview", "body"]

    public static func viewText(_ file: IndexedFile, view: String) -> String {
        switch view {
        case "identity":
            let stem = file.path.deletingPathExtension().lastPathComponent
            return Py.prefix(Py.collapseWhitespace("\(stem) \(file.title)"), 400)
        case "overview":
            return Py.prefix(Py.collapseWhitespace("\(file.summary) \(file.keywords.joined(separator: ", "))"), 600)
        case "body":
            return sample(file.text, budget: 1400)
        default: return ""
        }
    }

    static func distinctViews(_ file: IndexedFile) -> [(String, String)] {
        var seen: Set<String> = []
        return views.compactMap { view in
            let value = viewText(file, view: view)
            let key = Py.lower(value)
            guard !value.isEmpty, !seen.contains(key) else { return nil }
            seen.insert(key)
            return (view, value)
        }
    }

    /// Front/middle/tail sampling so long documents stay within the budget.
    static func sample(_ value: String, budget: Int) -> String {
        let collapsed = Py.collapseWhitespace(value)
        if Py.count(collapsed) <= budget { return collapsed }
        if budget < 12 { return Py.prefix(collapsed, budget) }
        let separatorBudget = Py.count(" … ") * 2
        let part = max(1, (budget - separatorBudget) / 3)
        let middle = max(0, Py.count(collapsed) / 2 - part / 2)
        let head = Py.prefix(collapsed, part)
        let centre = Py.slice(collapsed, middle, middle + part)
        let tail = Py.suffix(collapsed, part)
        return [head, centre, tail].joined(separator: " … ")
    }

    /// Short, stable representation used by both embedding and translation.
    public static func build(_ file: IndexedFile, maxChars: Int = 2400) -> String {
        precondition(maxChars >= 256, "semantic text budget must be at least 256 characters")

        var sections: [String] = []
        var seen: Set<String> = []
        let title = file.title.isEmpty ? file.path.deletingPathExtension().lastPathComponent : file.title
        for (label, value) in [("Title", title), ("Summary", file.summary), ("Keywords", file.keywords.joined(separator: ", "))] {
            let cleaned = Py.collapseWhitespace(value)
            let key = Py.lower(cleaned)
            if !cleaned.isEmpty && !seen.contains(key) {
                sections.append("\(label): \(cleaned)")
                seen.insert(key)
            }
        }
        let prefix = sections.joined(separator: "\n")
        let bodyBudget = max(0, maxChars - Py.count(prefix) - (prefix.isEmpty ? 0 : 2))
        let body = bodyBudget > 0 ? sample(file.text, budget: bodyBudget) : ""
        let parts = [prefix, body].filter { !$0.isEmpty }
        return Py.prefix(parts.joined(separator: "\n\n"), maxChars)
    }

    /// Choose a stable native embedding space from human-facing metadata.
    ///
    /// Code-heavy course notes frequently confuse automatic language detection.
    /// Titles and summaries are a better signal for the document's prose language.
    public static func language(_ file: IndexedFile) -> String {
        let title = file.title.isEmpty ? file.path.deletingPathExtension().lastPathComponent : file.title
        let sample = "\(title) \(Py.prefix(file.summary, 600))"
        let scalars = sample.unicodeScalars
        // The reference checks Hangul before kana regardless of position.
        if scalars.contains(where: { $0.value >= 0xAC00 && $0.value <= 0xD7AF }) { return "ko" }
        if scalars.contains(where: { $0.value >= 0x3040 && $0.value <= 0x30FF }) { return "ja" }
        var cjk = 0
        var latin = 0
        for scalar in scalars {
            if scalar.value >= 0x4E00 && scalar.value <= 0x9FFF { cjk += 1 }
            if scalar.isASCII && CharacterSet.letters.contains(scalar) { latin += 1 }
        }
        if cjk >= 2 && Double(cjk) / Double(max(1, cjk + latin)) >= 0.12 { return "zh-Hans" }
        return "en"
    }

    public static func cacheVersion(_ encoderVersion: String) -> String {
        "\(encoderVersion);semantic-text:\(version)"
    }
}
